src/runtime/refiner.ts

```diff
@@ -1237,143 +1241,220 @@ export class Fraction {
+      ordered[key] = unordered[key];
+    });
+    this._terms = ordered;
+    return this._terms;
+  }
+
+  set terms(tms: Dictionary<number>) {
+    this._terms = {...tms};
   }
 
-  static add(a: Polynomial, b: Polynomial): Polynomial {
-    Polynomial.assertCompatibility(a, b);
-    const sum = a.degree() > b.degree() ? Polynomial.copyOf(a) : Polynomial.copyOf(b);
-    const other = a.degree() > b.degree() ? b : a;
-    for (const [i, coeff] of other.coeffs.entries()) {
-      sum.coeffs[i] += coeff;
+  static add(a: Multinomial, b: Multinomial): Multinomial {
+    const sum = new Multinomial();
+    for (const [term, coeff] of Object.entries(a.terms)) {
+      const val = coeff + (sum.terms.hasOwnProperty(term) ? sum.terms[term] : 0);
+      sum.terms[term] = val;
+    }
+    for (const [term, coeff] of Object.entries(b.terms)) {
+      const val = coeff + (sum.terms.hasOwnProperty(term) ? sum.terms[term] : 0);
+      sum.terms[term] = val;
     }
     return sum;
   }
 
-  static subtract(a: Polynomial, b: Polynomial): Polynomial {
-    Polynomial.assertCompatibility(a, b);
-    return Polynomial.add(a, Polynomial.negate(b));
+  static subtract(a: Multinomial, b: Multinomial): Multinomial {
+    return Multinomial.add(a, Multinomial.negate(b));
   }
 
-  static negate(a: Polynomial): Polynomial {
-    return new Polynomial(a.coeffs.map(co => -co), a.indeterminate);
+  static negate(a: Multinomial): Multinomial {
+    const neg = Multinomial.copyOf(a);
+    for (const [term, coeff] of Object.entries(neg.terms)) {
+      neg.terms[term] = -coeff;
+    }
+    return neg;
   }
 
-  static multiply(a: Polynomial, b: Polynomial): Polynomial {
-    Polynomial.assertCompatibility(a, b);
-    const deg = a.degree() + b.degree();
-    const coeffs = new Array(deg+1).fill(0);
-    for (let i = 0; i < coeffs.length; i += 1) {
-      for (let j = 0; j <= i; j += 1) {
-        if (j <= a.degree() && (i-j) <= b.degree()) {
-          coeffs[i] += a.coeffs[j]*b.coeffs[i-j];
+  static multiply(a: Multinomial, b: Multinomial): Multinomial {
+    const prod = new Multinomial();
+    for (const [aKey, acoeff] of Object.entries(a.terms)) {
+      for (const [bKey, bcoeff] of Object.entries(b.terms)) {
+        const tprod = Term.fromKey(aKey);
+        const bterm = Term.fromKey(bKey);
+        for (const [indeterminate, power] of Object.entries(bterm.indeterminates)) {
+          const val = power + (tprod.indeterminates.hasOwnProperty(indeterminate) ? tprod.indeterminates[indeterminate] : 0);
+          tprod.indeterminates[indeterminate] = val;
         }
+        const val = acoeff*bcoeff + (prod.terms.hasOwnProperty(tprod.toKey()) ? prod.terms[tprod.toKey()] : 0);
+        prod.terms[tprod.toKey()] = val;
       }
     }
-    return new Polynomial(coeffs, a.indeterminate || b.indeterminate);
+    return prod;
   }
 
   isZero(): boolean {
-    return this.isConstant() && this.coeffs[0] === 0;
+    return Object.keys(this.terms).length === 0;
```

Comment:
I don't think this will work after two multinomials add or multiply and cancel out (as the add and multiply functions don't delete their 0 terms after building their result and this one doesn't ignore or remove them).

Reply:
The getter removes the 0s so I think it should work.

Reply:
Ah sorry, had thought that the getter was on a property of Term, nice!